bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt

```diff
@@ -0,0 +1,112 @@
+import org.eclipse.tractusx.bpdm.gate.api.model.response.SharingStateDto
+import org.eclipse.tractusx.bpdm.gate.entity.ChangelogEntry
+import org.eclipse.tractusx.bpdm.gate.entity.generic.BusinessPartner
+import org.eclipse.tractusx.bpdm.gate.repository.ChangelogRepository
+import org.eclipse.tractusx.bpdm.gate.repository.generic.BusinessPartnerRepository
+import org.springframework.data.domain.PageRequest
+import org.springframework.stereotype.Service
+
+@Service
+class BusinessPartnerService(
+    private val businessPartnerRepository: BusinessPartnerRepository,
+    private val businessPartnerMappings: BusinessPartnerMappings,
+    private val sharingStateService: SharingStateService,
+    private val changelogRepository: ChangelogRepository
+) {
+
+    fun upsertBusinessPartnersInput(dtos: Collection<BusinessPartnerInputRequest>): Collection<BusinessPartnerInputDto> {
+        val existingEntitiesByExternalId = getExistingEntityByExternalId(dtos, StageType.Input)
+
+        val savedEntities = dtos.map { dto ->
+            existingEntitiesByExternalId[dto.externalId]
+                ?.let { existingEntity -> updateBusinessPartnerInput(existingEntity, dto) }
+                ?: run { insertBusinessPartnerInput(dto) }
+        }
+
+        return savedEntities.map(businessPartnerMappings::toBusinessPartnerInputDto)
+    }
+
+    private fun insertBusinessPartnerInput(dto: BusinessPartnerInputRequest): BusinessPartner {
+        val newEntity = businessPartnerMappings.toBusinessPartner(dto, StageType.Input)
+        return businessPartnerRepository.save(newEntity)
+            .also {
+                saveChangelog(dto.externalId, ChangelogType.CREATE, StageType.Input)
+                initSharingState(dto)
+            }
+    }
+
+    private fun updateBusinessPartnerInput(existingEntity: BusinessPartner, dto: BusinessPartnerInputRequest): BusinessPartner {
+        businessPartnerMappings.updateBusinessPartner(existingEntity, dto)
+        return businessPartnerRepository.save(existingEntity)
+            .also {
+                saveChangelog(dto.externalId, ChangelogType.UPDATE, StageType.Input)
+            }
+    }
+
+    fun getBusinessPartnersInput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerInputDto> {
+        val stage = StageType.Input
+        val page = when {
+            externalIds != null -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
+            else -> businessPartnerRepository.findByStage(stage, pageRequest)
+        }
+        return page.toPageDto(businessPartnerMappings::toBusinessPartnerInputDto)
+    }
+
+    fun getBusinessPartnersOutput(pageRequest: PageRequest, externalIds: Collection<String>?): PageDto<BusinessPartnerOutputDto> {
+        val stage = StageType.Output
+        val page = when {
+            externalIds != null -> businessPartnerRepository.findByStageAndExternalIdIn(stage, externalIds, pageRequest)
```

Comment:
Behaviour in the Gate seems to be inconsistent regarding missing externalIDs. I would recommend to take the same behaviour as in the Pool: When the list of IDs is null or *empty* then return all objects.

Reply:
From a logical point of view handling null and empty collections differently makes a little more sense to me (null means no filtering; empty means I want only these specific business partners but there are none).
But I agree, from a practical perspective, the empty array case doesn't make a lot of sense and it makes the API probably easier to use, if null and empty are handled the same.
So I changed the check to isNullOrEmpty().